File: frontend/src/api/campaigns.js
const API_BASE_URL =process.env.REACT_APP_API_BASE_URL || "http://backend:5001/api";
export async function createCampaign(data) {
  const response = await fetch(`${API_BASE_URL}/campaigns`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to create campaign");
  }

  return response.json();
}

export async function getCampaigns() {
  const response = await fetch(`${API_BASE_URL}/campaigns`);

  if (!response.ok) {
    throw new Error("Failed to fetch campaigns");
  }

  return response.json();
}

export async function publishCampaign(campaignId) {
  const response = await fetch(
    `${API_BASE_URL}/campaigns/${campaignId}/publish`,
    { method: "POST" }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to publish campaign");
  }

  return response.json();
}

export async function pauseCampaign(campaignId) {
  const response = await fetch(
    `${API_BASE_URL}/campaigns/${campaignId}/pause`,
    { method: "POST" }
  );

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "Failed to pause campaign");
  }

  return response.json();
}

File: backend/app/routes/campaigns.py
from flask import Blueprint, request, jsonify
from app.services.campaign_service import CampaignService
from app.errors.exceptions import ValidationError

campaigns_bp = Blueprint("campaigns", __name__, url_prefix="/api/campaigns")


@campaigns_bp.route("", methods=["POST"])
def create_campaign():
    data = request.get_json()

    if not data:
        raise ValidationError("Invalid JSON body")

    campaign = CampaignService.create_campaign(data)

    return jsonify({
        "id": str(campaign.id),
        "status": campaign.status
    }), 201


@campaigns_bp.route("", methods=["GET"])
def list_campaigns():
    campaigns = CampaignService.list_campaigns()

    return jsonify([
        {
            "id": str(c.id),
            "name": c.name,
            "objective": c.objective,
            "campaign_type": c.campaign_type,
            "daily_budget": c.daily_budget,
            "status": c.status,
            "google_campaign_id": c.google_campaign_id,
            "created_at": c.created_at.isoformat()
        }
        for c in campaigns
    ]), 200


@campaigns_bp.route("/<uuid:campaign_id>/publish", methods=["POST"])
def publish_campaign(campaign_id):
    campaign = CampaignService.publish_campaign(campaign_id)

    return jsonify({
        "id": str(campaign.id),
        "status": campaign.status,
        "google_campaign_id": campaign.google_campaign_id
    }), 200


@campaigns_bp.route("/<uuid:campaign_id>/pause", methods=["POST"])
def pause_campaign(campaign_id):
    campaign = CampaignService.pause_campaign(campaign_id)

    return jsonify({
        "id": str(campaign.id),
        "status": campaign.status
    }), 200